src/server/session.ts

```diff
@@ -1266,28 +1277,81 @@ export class Session<TMessage = string> implements EventSender {
         tracing?.push(tracing.Phase.Session, "suggestionCheck", { file, configFilePath: (project as ConfiguredProject).canonicalConfigFilePath }); // undefined is fine if the cast fails
         this.sendDiagnosticsEvent(file, project, project.getLanguageService().getSuggestionDiagnostics(file), "suggestionDiag");
         tracing?.pop();
     }
 
-    private sendDiagnosticsEvent(file: NormalizedPath, project: Project, diagnostics: readonly Diagnostic[], kind: protocol.DiagnosticEventKind): void {
+    private regionSemanticCheck(file: NormalizedPath, project: Project, ranges: TextRange[]): boolean {
+        this.diagnosticsTime = this.hrtime();
+        tracing?.push(tracing.Phase.Session, "regionSemanticCheck", { file, configFilePath: (project as ConfiguredProject).canonicalConfigFilePath }); // undefined is fine if the cast fails
+        let diagnosticsResult;
+        if (!this.shouldDoRegionCheck(file) || !(diagnosticsResult = project.getLanguageService().getRegionSemanticDiagnostics(file, ranges))) {
+            tracing?.pop();
+            return false;
+        }
+        this.sendDiagnosticsEvent(file, project, diagnosticsResult.diagnostics, "regionSemanticDiag", diagnosticsResult.spans);
+        tracing?.pop();
+        return true;
+    }
+
+    // We should only do the region-based semantic check if we think it would be considerably faster than a whole-file semantic check
+    /** @internal */
+    protected shouldDoRegionCheck(file: NormalizedPath): boolean {
+        const lineCount = this.projectService.getScriptInfoForNormalizedPath(file)?.getLineCount();
+        const perf = this.semanticCheckPerformance.get(file);
+        return !!(lineCount && lineCount > 500 && (!perf || perf > 200));
+    }
+
+    /** @internal */
+    private cleanUpSemanticCheckPerformance(closedFiles: string[] | undefined): void {
+        if (closedFiles) {
+            closedFiles.forEach(file => this.semanticCheckPerformance.delete(toNormalizedPath(file)));
+        }
+    }
+
+    private sendDiagnosticsEvent(file: NormalizedPath, project: Project, diagnostics: readonly Diagnostic[], kind: protocol.DiagnosticEventKind, spans?: TextSpan[]): void {
         try {
-            this.event<protocol.DiagnosticEventBody>({ file, diagnostics: diagnostics.map(diag => formatDiag(file, project, diag)) }, kind);
+            const scriptInfo = Debug.checkDefined(project.getScriptInfo(file));
+            const duration = hrTimeToMilliseconds(this.hrtime(this.diagnosticsTime));
+            if (kind === "semanticDiag" && !this.shouldDoRegionCheck(file)) {
+                this.semanticCheckPerformance.set(file, duration);
+            }
+
+            const body: protocol.DiagnosticEventBody = {
+                file,
+                diagnostics: diagnostics.map(diag => formatDiag(file, project, diag)),
+                spans: spans?.map(span => toProtocolTextSpan(span, scriptInfo)),
+                duration,

```

Comment:
Instead of passing this here do we want to use existing machinery we use for performance events that is : 
- Do we need this on every event per file per duiagnostic kind or can it be : on diagnostic completion event .
- Name this similar to out `protocol.Respose` as `performanceData` with correct keys

Reply:
Also pls check if VS should be ok with this extra data coming in the event. I have completely lost track of if and when they use the events .. i think they use that for project diagnostics (GeterrForProjectRequest)

Reply:
I think having the duration per file and diagnostic kind will help analyze the data we will collect via vscode telemetry. Is there a significant downside to having this data per file per kind of diagnostics?